akka-http/src/main/scala/akka/http/scaladsl/server/JsonEntityStreaming.scala

```diff
@@ -0,0 +1,169 @@
+package akka.http.scaladsl.server
+
+import akka.actor.ActorSystem
+import akka.http.impl.util.JavaMapping
+import akka.http.scaladsl.model.{ ContentType, ContentTypes }
+import akka.http.scaladsl.server.directives.EntityStreamingDirectives.SourceRenderingMode
+import akka.stream.io.Framing
+import akka.stream.scaladsl.Flow
+import akka.util.ByteString
+import com.typesafe.config.Config
+
+import scala.collection.immutable
+
+/**
+ * Same as [[Framing]] but additionally can express which [[ContentType]] it supports,
+ * which can be used to reject routes if content type does not match used framing.
+ */
+abstract class FramingWithContentType extends Framing { // TODO javadsl
+  def flow: Flow[ByteString, ByteString, Unit]
+  def supported: immutable.Set[ContentType]
+  def supports(ct: ContentType): Boolean = supported.contains(ct)
+  def isSupported(ct: akka.http.javadsl.model.ContentType): Boolean = supports(JavaMapping.ContentType.toScala(ct))
+}
+object FramingWithContentType {
+  def apply(framing: Flow[ByteString, ByteString, Any], contentType: ContentType, moreContentTypes: ContentType*) =
+    new FramingWithContentType {
+      // TODO javadsl
+      override def flow: Flow[ByteString, ByteString, Unit] = framing.mapMaterializedValue(_ ⇒ ())
+
+      override def supported: Set[ContentType] =
+        if (moreContentTypes.isEmpty) Set(contentType)
+        else Set(contentType) ++ moreContentTypes
+    }
+}
+
+trait JsonEntityStreamingSupport {
+
+  /** `application/json` specific Framing implementation */
+  implicit object JsonFraming extends FramingWithContentType { // TODO
+    override final val flow = Flow[ByteString].via(Framing.json(Int.MaxValue))
+
+    override def supports(ct: ContentType): Boolean = ct match {
+      case ContentTypes.`application/json` ⇒ true
+      case _                               ⇒ false
+    }
+
+    final val _supported = Set(ContentTypes.`application/json`)
+
+    override def supported: Set[ContentType] = _supported
+  }
+
+}
+
+/**
+ * Specialised rendering mode for streaming elements as JSON.
+ *
+ * See also: <a href="https://en.wikipedia.org/wiki/JSON_Streaming">JSON Streaming on Wikipedia</a>.
+ */
+trait JsonSourceRenderingMode extends SourceRenderingMode
+
+// TODO framing needs a max-entity size.
+object JsonSourceRenderingMode {
```

Comment:
Not sure if this matters, but won't the `override def between = ByteString("c")` below allocate a new identical object for each chunk?